File: src/main/java/com/wesker926/leetcode/algorithms/p0301/s1/Solution.java
package com.wesker926.leetcode.algorithms.p0301.s1;

import java.util.LinkedList;
import java.util.List;

/**
 * @author wesker.gh
 * @date 2021/10/27
 * @description dfs
 */
public class Solution {
    public List<String> removeInvalidParentheses(String s) {
        int lr = 0, rr = 0;
        for (char c : s.toCharArray()) {
            if (c == '(') {
                lr++;
            } else if (c == ')') {
                if (lr == 0) {
                    rr++;
                } else {
                    lr--;
                }
            }
        }

        List<String> result = new LinkedList<>();
        dfs(result, s, 0, 0, 0, lr, rr);
        return result;
    }

    private void dfs(List<String> result, String str, int start, int lc, int rc, int lr, int rr) {
        if (lr == 0 && rr == 0) {
            if (valid(str)) {
                result.add(str);
            }
            return;
        }

        char[] cs = str.toCharArray();
        for (int i = start; i < cs.length; i++) {
            if (cs[i] != '(' && cs[i] != ')') {
                continue;
            }

            if (cs.length - i < lr + rr) {
                return;
            }

            boolean left = cs[i] == '(';
            if (i == start || cs[i] != cs[i - 1]) {
                String s = new String(cs, 0, i) + new String(cs, i + 1, cs.length - i - 1);
                if (lr > 0 && left) {
                    dfs(result, s, i, lc, rc, lr - 1, rr);
                }

                if (rr > 0 && !left) {
                    dfs(result, s, i, lc, rc, lr, rr - 1);
                }
            }

            lc = left ? lc + 1 : lc;
            rc = left ? rc : rc + 1;
            if (rc > lc) {
                return;
            }
        }
    }

    private boolean valid(String str) {
        int count = 0;
        for (char c : str.toCharArray()) {
            if (c == '(') {
                count++;
            } else if (c == ')') {
                count--;
                if (count < 0) {
                    return false;
                }
            }
        }
        return count == 0;
    }
}

File: src/main/java/com/wesker926/leetcode/algorithms/p0301/s1/solution.go
// Package s1
// @author: wesker.gh
// @date: 2021/10/27
// @description: dfs
package s1

var result []string

func removeInvalidParentheses(s string) []string {
	lr, rr := 0, 0
	for i := 0; i < len(s); i++ {
		if s[i] == '(' {
			lr++
		} else if s[i] == ')' {
			if lr == 0 {
				rr++
			} else {
				lr--
			}
		}
	}

	result = make([]string, 0)
	dfs(s, 0, 0, 0, lr, rr)
	return result
}

func dfs(str string, start, lc, rc, lr, rr int) {
	if lr == 0 && rr == 0 {
		if valid(str) {
			result = append(result, str)
		}
		return
	}

	var r bool
	for i := start; i < len(str); i++ {
		if str[i] != '(' && str[i] != ')' {
			continue
		}

		if len(str)-i < lr+rr {
			return
		}

		left := str[i] == '('
		if i == start || str[i] != str[i-1] {
			if lr > 0 && left {
				dfs(str[:i]+str[i+1:], i, lc, rc, lr-1, rr)
			}

			if rr > 0 && !left {
				dfs(str[:i]+str[i+1:], i, lc, rc, lr, rr-1)
			}
		}

		if lc, rc, r = countAndCut(str[i], lc, rc); !r {
			return
		}
	}
}

func valid(str string) bool {
	count := 0
	for i := 0; i < len(str); i++ {
		if str[i] == '(' {
			count++
		} else if str[i] == ')' {
			count--
			if count < 0 {
				return false
			}
		}
	}

	if count == 0 {
		return true
	}
	return false
}

func countAndCut(c byte, lc, rc int) (int, int, bool) {
	if c == '(' {
		lc++
	} else {
		rc++
	}

	r := rc <= lc
	return lc, rc, r
}
